### ISSUE TITLE
Warn on unnecessary coercions

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust-clippy#3526]
Here the `as *const i32` coercion is unnecessary, but no warning is emitted.

```rust
fn foo(_x: *const i32) {}

pub fn bar(x: &mut i32) {
    foo(x as *const i32)
}

pub fn baz(x: &mut i32) {
    foo(x)
}
```

cc @oli-obk
